In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
df = pd.read_excel(r'C:\Users\ycc\Downloads\MaxAB Task\dataset1.xlsx')

In [3]:
df.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  
0  Application     العمرانية                      2119.75      0.00  
1  Application      ميت عقبة                      2242.75      0.10  
2  Application  مدينة السلام                      1094.50      8.65  
3  Application      ميت عقبة                      3448.50     24.51  
4  Application       المطرية                       910.95      0.10

In [4]:
df_copy = df.copy()

In [5]:
df_copy.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  
0  Application     العمرانية                      2119.75      0.00  
1  Application      ميت عقبة                      2242.75      0.10  
2  Application  مدينة السلام                      1094.50      8.65  
3  Application      ميت عقبة                      3448.50     24.51  
4  Application       المطرية                       910.95      0.10

In [6]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21716 entries, 0 to 21715
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   week                         21716 non-null  int64  
 1   month                        21716 non-null  int64  
 2   creation_date                21716 non-null  object 
 3   sales_order_id               21716 non-null  int64  
 4   retailer_id                  21716 non-null  int64  
 5   sales_order_status           21716 non-null  object 
 6   channel                      21716 non-null  object 
 7   district_ar                  21716 non-null  object 
 8   order_price_before_discount  21716 non-null  float64
 9   discount                     21716 non-null  float64
dtypes: float64(2), int64(4), object(4)
memory usage: 1.7+ MB


In [7]:
df_copy.isnull().sum()

week                           0
month                          0
creation_date                  0
sales_order_id                 0
retailer_id                    0
sales_order_status             0
channel                        0
district_ar                    0
order_price_before_discount    0
discount                       0
dtype: int64

In [8]:
df_copy.duplicated().sum()

0

In [9]:
df_copy.head(10)

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   
5    14      4    2020-04-01              21        25782           Canceled   
6    14      4    2020-04-01              25        10497             Failed   
7    14      4    2020-04-01              29        17783          Delivered   
8    14      4    2020-04-01              33        10025          Delivered   
9    14      4    2020-04-01              37         3027          Delivered   

       channel   district_ar  order_price_before_discount  discount  
0  Application     العمرانية                      2119.75      0.00  
1  Application      ميت عقبة                      2242.75      0.10  
2  Application  مدينة السلام                      1094.50      8.65  
3  Application      ميت عقبة                      3448.50     24.51  
4  Application       المطرية                       910.95      0.10  
5  Application     صفط اللبن                     27518.50    236.59  
6  Application        الوراق                       806.75      0.00  
7  Application         هرم 3                      2113.45      0.10  
8  Application       عين شمس                      3369.75     35.98  
9  Application       المطرية                      1365.90      0.10

# Question 1

### part 1: may retenstion rate

In [10]:
df_april = df_copy[df_copy['month'] == 4]

In [11]:
april_retailers = df_april['retailer_id'].unique().tolist()

In [12]:
len(april_retailers)

5004

we have 5004 retailer in april, we want to know how many of the 5004 stay with us on may

In [13]:
df_april_retailers = df_april.drop_duplicates(subset = 'retailer_id')

In [14]:
df_april_retailers

week  month creation_date  sales_order_id  retailer_id  \
0       14      4    2020-04-01               1        18564   
1       14      4    2020-04-01               5        28140   
2       14      4    2020-04-01               9        32552   
3       14      4    2020-04-01              13        19299   
4       14      4    2020-04-01              17         5901   
...    ...    ...           ...             ...          ...   
8645    18      4    2020-04-30           34583        25273   
8646    18      4    2020-04-30           34587         7366   
8653    18      4    2020-04-30           34615        18852   
8658    18      4    2020-04-30           34635        22677   
8659    18      4    2020-04-30           34639        35525   

     sales_order_status      channel      district_ar  \
0             Delivered  Application        العمرانية   
1             Delivered  Application         ميت عقبة   
2              Canceled  Application     مدينة السلام   
3             Delivered  Application         ميت عقبة   
4             Delivered  Application          المطرية   
...                 ...          ...              ...   
8645          Delivered  Application            المرج   
8646          Delivered  Application  الزاوية الحمراء   
8653          Delivered  Application            المرج   
8658          Delivered  Application             فيصل   
8659          Delivered  Application          الزيتون   

      order_price_before_discount  discount  
0                         2119.75      0.00  
1                         2242.75      0.10  
2                         1094.50      8.65  
3                         3448.50     24.51  
4                          910.95      0.10  
...                           ...       ...  
8645                      3936.10     59.22  
8646                      1271.00      7.90  
8653                       922.50      0.00  
8658                      3244.00     50.18  
8659                      2474.75     15.79  

[5004 rows x 10 columns]

In [15]:
# may retailers

df_may = df_copy[df_copy['month'] == 5]
retained_from_april_to_may = df_may['retailer_id'].isin(april_retailers)

In [16]:
retained_from_april_to_may_count = retained_from_april_to_may.value_counts()[1]

In [17]:
may_retention_rate = (retained_from_april_to_may_count / len(april_retailers))*100

In [18]:
may_retention_rate

90.24780175859313

### part 2: retention rate for each channel

In [19]:
df_copy['channel'].value_counts()

Application    17072
salesagent      2750
telesales       1894
Name: channel, dtype: int64

In [20]:
application_april = df_april[df_april['channel'] == 'Application']['retailer_id'].unique().tolist()
salesagent_april = df_april[df_april['channel'] == 'salesagent']['retailer_id'].unique().tolist()
telesales_april = df_april[df_april['channel'] == 'telesales']['retailer_id'].unique().tolist()

In [21]:
retained_from_april_to_may_application = df_may['retailer_id'].isin(application_april)
retained_from_april_to_may_salesagent = df_may['retailer_id'].isin(salesagent_april)
retained_from_april_to_may_telesales = df_may['retailer_id'].isin(telesales_april)

In [22]:
retained_from_april_to_may_application_count = retained_from_april_to_may_application.value_counts()[1]
retained_from_april_to_may_salesagent_count = retained_from_april_to_may_salesagent.value_counts()[1]
retained_from_april_to_may_telesales_count = retained_from_april_to_may_telesales.value_counts()[1]

In [23]:
retained_from_april_to_may_application_count

4060

In [24]:
retained_from_april_to_may_salesagent_count

246

In [25]:
retained_from_april_to_may_telesales_count

408

In [26]:
retention_application_rate = (retained_from_april_to_may_application_count / len(application_april))*100
retention_salesagent_rate = (retained_from_april_to_may_salesagent_count / len(salesagent_april))*100
retention_telesales_rate = (retained_from_april_to_may_telesales_count / len(telesales_april))*100

In [27]:
retention_application_rate

99.75429975429975

In [28]:
retention_salesagent_rate

49.2

In [29]:
retention_telesales_rate

70.58823529411765

# Question 2:

### part 1 : % of growth of may ove april for every district in terms of per number of orders

In [30]:
df_copy['month'].value_counts()

4    8661
5    8257
6    4798
Name: month, dtype: int64

In [31]:
df_AprilAndMay = df_copy[df_copy['month'] != 6 ]
df_AprilAndMay['month'].value_counts()

4    8661
5    8257
Name: month, dtype: int64

In [32]:
# Order Growth Rate = (# of Orders in May - # of Orders in April) / # of Orders in April

df_april['district_ar'].count()

8661

In [33]:
df_may['district_ar'].count()

8257

In [34]:
len(df_AprilAndMay['district_ar'].unique())

78

In [35]:
number_of_orders_for_each_district = df_AprilAndMay.groupby(['month','district_ar']).size().unstack()
number_of_orders_for_each_district.dropna(axis = 1, inplace = True)

In [36]:
number_of_orders_for_each_district

district_ar  أرض اللواء  البراجيل  الجبل الاصفر  الخانكة  الخصوص  الدقي  \
month                                                                     
4                 205.0      79.0          54.0      9.0   116.0   93.0   
5                 189.0      75.0          47.0     14.0   140.0   77.0   

district_ar  الزاوية الحمراء  الزيتون  السبتية  الشرابية  ...  مدينة الحرفيين  \
month                                                     ...                   
4                      137.0    173.0     68.0     130.0  ...           102.0   
5                      116.0    148.0     68.0      91.0  ...           105.0   

district_ar  مدينة السلام  مساكن شيراتون  مسطرد  منطي  ميت حلفا  ميت عقبة  \
month                                                                       
4                   122.0           15.0   87.0  21.0      10.0      62.0   
5                    86.0            5.0   70.0  19.0       6.0      71.0   

district_ar  ميت نما  هرم 2  هرم 3  
month                               
4                9.0  292.0  267.0  
5                9.0  270.0  209.0  

[2 rows x 64 columns]

In [37]:
number_of_orders_for_each_district.shape

(2, 64)

In [38]:
orderGrowthDistrict = pd.DataFrame(index =number_of_orders_for_each_district.columns, columns = [4,5,'percentageOf_GrowthInNumOfOrders'])

In [39]:
orderGrowthDistrict[4] = number_of_orders_for_each_district.loc[4, :]
orderGrowthDistrict[5] = number_of_orders_for_each_district.loc[5, :]
orderGrowthDistrict['percentageOf_GrowthInNumOfOrders'] = np.round((((number_of_orders_for_each_district.loc[5,] - number_of_orders_for_each_district.loc[4,]) / number_of_orders_for_each_district.loc[5,])) * 100,1)

In [40]:
orderGrowthDistrict.sort_values('percentageOf_GrowthInNumOfOrders' ,ascending = False)

4     5  percentageOf_GrowthInNumOfOrders
district_ar                                                
شبين القناطر    1.0   3.0                              66.7
القطاوي         1.0   2.0                              50.0
بلقس            9.0  17.0                              47.1
قلج            27.0  48.0                              43.8
الخانكة         9.0  14.0                              35.7
...             ...   ...                               ...
بهتيم          53.0  32.0                             -65.6
ميت حلفا       10.0   6.0                             -66.7
الوايلي        61.0  33.0                             -84.8
مساكن شيراتون  15.0   5.0                            -200.0
عمرانيه 3      16.0   4.0                            -300.0

[64 rows x 3 columns]

In [41]:
orderGrowthDistrict.loc['شبين القناطر']

4                                    1.0
5                                    3.0
percentageOf_GrowthInNumOfOrders    66.7
Name: شبين القناطر, dtype: float64

In [42]:
orderGrowthDistrict.loc['عمرانيه 3']

4                                    16.0
5                                     4.0
percentageOf_GrowthInNumOfOrders   -300.0
Name: عمرانيه 3, dtype: float64

In [43]:
df_AprilAndMay[(df_AprilAndMay['month'] == 4) & (df_AprilAndMay['district_ar'] == 'أبو رواش')]

week  month creation_date  sales_order_id  retailer_id  \
187     14      4    2020-04-01             749        27322   
5029    17      4    2020-04-18           20117        27322   

     sales_order_status    channel district_ar  order_price_before_discount  \
187           Delivered  telesales    أبو رواش                      9034.75   
5029          Delivered  telesales    أبو رواش                      3373.05   

      discount  
187      59.40  
5029     46.28

In [44]:
df_AprilAndMay[(df_AprilAndMay['month'] == 4) & (df_AprilAndMay['district_ar'] == 'الخانكة')]

week  month creation_date  sales_order_id  retailer_id  \
313     14      4    2020-04-02            1253        31321   
4541    16      4    2020-04-16           18165        32297   
5796    17      4    2020-04-21           23187        31321   
7360    18      4    2020-04-25           29443        32297   
7361    18      4    2020-04-25           29447        32297   
7363    18      4    2020-04-25           29455        32297   
7365    18      4    2020-04-25           29463        32297   
7991    18      4    2020-04-27           31967        32297   
8181    18      4    2020-04-28           32727        31298   

     sales_order_status      channel district_ar  order_price_before_discount  \
313            Canceled  Application     الخانكة                       1718.1   
4541          Delivered  Application     الخانكة                      28655.0   
5796          Delivered  Application     الخانكة                       4737.0   
7360           Canceled  Application     الخانكة                       6889.0   
7361           Canceled  Application     الخانكة                       6889.0   
7363           Canceled  Application     الخانكة                       6889.0   
7365           Canceled  Application     الخانكة                       6889.0   
7991          Delivered  Application     الخانكة                       4464.5   
8181          Delivered   salesagent     الخانكة                       1730.5   

      discount  
313      17.18  
4541    525.33  
5796     14.90  
7360     37.70  
7361     37.70  
7363     37.70  
7365     37.70  
7991     89.29  
8181      0.00

### part 2 : % of growth of may ove april for every district in terms of number of distinct retailers

In [45]:
DistinctRetailers = pd.pivot_table(data=df_AprilAndMay, values= 'retailer_id',index= 'month', columns= 'district_ar',aggfunc= pd.Series.nunique) 
DistinctRetailers.dropna(axis = 1, inplace = True)

In [46]:
DistinctRetailers

district_ar  أرض اللواء  البراجيل  الجبل الاصفر  الخانكة  الخصوص  الدقي  \
month                                                                     
4                 115.0      53.0          31.0      3.0    76.0   47.0   
5                 121.0      50.0          32.0      6.0    93.0   43.0   

district_ar  الزاوية الحمراء  الزيتون  السبتية  الشرابية  ...  مدينة الحرفيين  \
month                                                     ...                   
4                       83.0     99.0     36.0      84.0  ...            59.0   
5                       70.0     87.0     39.0      57.0  ...            64.0   

district_ar  مدينة السلام  مساكن شيراتون  مسطرد  منطي  ميت حلفا  ميت عقبة  \
month                                                                       
4                    75.0            8.0   48.0   9.0       5.0      31.0   
5                    57.0            5.0   49.0  16.0       5.0      45.0   

district_ar  ميت نما  هرم 2  هرم 3  
month                               
4                6.0  182.0  145.0  
5                5.0  178.0  129.0  

[2 rows x 64 columns]

In [47]:
GrowthInNumberOfDistinctRetailers = pd.DataFrame(index = DistinctRetailers.columns, columns = [4, 5, 'PercentageOf_GrowthInNumberOfDistinctRetailers'])

In [48]:
GrowthInNumberOfDistinctRetailers[4] = DistinctRetailers.loc[4, ]
GrowthInNumberOfDistinctRetailers[5] = DistinctRetailers.loc[5, ]
GrowthInNumberOfDistinctRetailers['PercentageOf_GrowthInNumberOfDistinctRetailers'] = np.round((((DistinctRetailers.loc[5, ] - DistinctRetailers.loc[4, ]) / DistinctRetailers.loc[5, ]) * 100),1) 

In [49]:
GrowthInNumberOfDistinctRetailers.value_counts()[1]

5    PercentageOf_GrowthInNumberOfDistinctRetailers
1.0  0.0                                               1
2.0  50.0                                              1
dtype: int64

In [50]:
GrowthInNumberOfDistinctRetailers.sort_values('PercentageOf_GrowthInNumberOfDistinctRetailers', ascending = False)

4      5  PercentageOf_GrowthInNumberOfDistinctRetailers
district_ar                                                                
القطاوي          1.0    2.0                                            50.0
قلج             15.0   30.0                                            50.0
الخانكة          3.0    6.0                                            50.0
منطي             9.0   16.0                                            43.8
بشتيل           73.0  117.0                                            37.6
...              ...    ...                                             ...
الوايلي         28.0   19.0                                           -47.4
الشرابية        84.0   57.0                                           -47.4
النزهة الجديدة  34.0   22.0                                           -54.5
مساكن شيراتون    8.0    5.0                                           -60.0
عمرانيه 3       11.0    4.0                                          -175.0

[64 rows x 3 columns]

In [51]:
GrowthInNumberOfDistinctRetailers.loc['منطي']

4                                                  9.0
5                                                 16.0
PercentageOf_GrowthInNumberOfDistinctRetailers    43.8
Name: منطي, dtype: float64

#### part 3:

In [52]:
df_AprilAndMay['order_price'] = df_AprilAndMay['order_price_before_discount'] - df_AprilAndMay['discount']

<ipython-input-52-bb7d801e6a0a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_AprilAndMay['order_price'] = df_AprilAndMay['order_price_before_discount'] - df_AprilAndMay['discount']


In [53]:
df_AprilAndMay.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  \
0  Application     العمرانية                      2119.75      0.00   
1  Application      ميت عقبة                      2242.75      0.10   
2  Application  مدينة السلام                      1094.50      8.65   
3  Application      ميت عقبة                      3448.50     24.51   
4  Application       المطرية                       910.95      0.10   

   order_price  
0      2119.75  
1      2242.65  
2      1085.85  
3      3423.99  
4       910.85

In [54]:
avg_order_price_pivot = pd.pivot_table(data = df_AprilAndMay, values = 'order_price', columns = 'district_ar', index = 'month', aggfunc = 'mean')
avg_order_price_pivot.dropna(axis = 1, inplace = True)

In [55]:
avg_order_price_pivot

district_ar   أرض اللواء     البراجيل  الجبل الاصفر      الخانكة       الخصوص  \
month                                                                           
4            3247.118496  2853.815443   3457.240463  7562.622222  3101.221609   
5            3007.027672  2406.137733   2929.732979  7109.197143  2595.444786   

district_ar        الدقي  الزاوية الحمراء      الزيتون      السبتية  \
month                                                                 
4            2670.976183      3336.849124  3214.798555  2934.494779   
5            2143.086104      2112.916638  3036.876757  3141.250294   

district_ar     الشرابية  ...  مدينة الحرفيين  مدينة السلام  مساكن شيراتون  \
month                     ...                                                
4            2944.686192  ...     3948.378284   3046.608087    3105.006667   
5            2522.849890  ...     2560.014476   2292.153837    1572.870000   

district_ar        مسطرد         منطي  ميت حلفا     ميت عقبة      ميت نما  \
month                                                                       
4            3226.342701  2789.270952  4203.408  2376.858548  2912.954444   
5            3063.117143  2622.103158  2023.115  2135.190423  2034.924444   

district_ar        هرم 2        هرم 3  
month                                  
4            2644.497363  3000.039594  
5            2558.080333  3061.576124  

[2 rows x 64 columns]

In [56]:
avg_order_price = pd.DataFrame(columns = [4,5,'AvgOrderPrice'], index = avg_order_price_pivot.columns )

In [57]:
avg_order_price[4] = avg_order_price_pivot.loc[4,]
avg_order_price[5] = avg_order_price_pivot.loc[5,]
avg_order_price['AvgOrderPrice'] = np.round((((avg_order_price_pivot.loc[5,] - avg_order_price_pivot.loc[4,]) / avg_order_price_pivot.loc[5,])*100),1)

In [58]:
avg_order_price.sort_values('AvgOrderPrice', ascending = False)

4            5  AvgOrderPrice
district_ar                                             
المطرية          2440.839011  2755.929390           11.4
شبين القناطر     1564.050000  1739.416667           10.1
القناطر الخيرية  2835.368562  3085.806341            8.1
شبرا مصر         2573.695040  2770.376794            7.1
بهتيم            3157.266038  3384.227188            6.7
...                      ...          ...            ...
الزاوية الحمراء  3336.849124  2112.916638          -57.9
مساكن شيراتون    3105.006667  1572.870000          -97.4
ميت حلفا         4203.408000  2023.115000         -107.8
عمرانيه 3        3038.448125  1274.722500         -138.4
القطاوي          3037.810000   842.360000         -260.6

[64 rows x 3 columns]

### part 4:

In [59]:
total_revenue_pivot = pd.pivot_table(data = df_AprilAndMay, values = 'order_price', columns = 'district_ar', index = 'month', aggfunc = 'sum')
total_revenue_pivot.dropna(axis = 1, inplace = True)

In [60]:
total_revenue_pivot

district_ar     أرض اللواء   البراجيل  الجبل الاصفر   الخانكة         الخصوص  \
month                                                                          
4            665659.291667  225451.42    186690.985  68063.60  359741.706667   
5            568328.230000  180460.33    137697.450  99528.76  363362.270000   

district_ar       الدقي  الزاوية الحمراء    الزيتون     السبتية    الشرابية  \
month                                                                         
4            248400.785        457148.33  556160.15  199545.645  382809.205   
5            165017.630        245098.33  449457.76  213605.020  229579.340   

district_ar  ...  مدينة الحرفيين   مدينة السلام  مساكن شيراتون       مسطرد  \
month        ...                                                             
4            ...      402734.585  371686.186667       46575.10  280691.815   
5            ...      268801.520  197125.230000        7864.35  214418.200   

district_ar      منطي  ميت حلفا   ميت عقبة   ميت نما      هرم 2          هرم 3  
month                                                                           
4            58574.69  42034.08  147365.23  26216.59  772193.23  801010.571667  
5            49819.96  12138.69  151598.52  18314.32  690681.69  639869.410000  

[2 rows x 64 columns]

In [61]:
total_revenue = pd.DataFrame(columns = [4,5,'TotalRevenues'], index = total_revenue_pivot.columns )

In [62]:
total_revenue[4] = total_revenue_pivot.loc[4,]
total_revenue[5] = total_revenue_pivot.loc[5,]
total_revenue['TotalRevenues'] = np.round((((total_revenue_pivot.loc[5,] - total_revenue_pivot.loc[4,]) / total_revenue_pivot.loc[5,])*100),1)

In [63]:
total_revenue.sort_values('TotalRevenues', ascending = False)

4          5  TotalRevenues
district_ar                                          
شبين القناطر       1564.050    5218.25           70.0
قلج               91354.400  141871.32           35.6
بلقس              30068.270   45182.40           33.5
الخانكة           68063.600   99528.76           31.6
القناطر الخيرية  206981.905  253036.12           18.2
...                     ...        ...            ...
الوايلي          199722.960   93521.89         -113.6
النزهة الجديدة   158944.570   64815.53         -145.2
ميت حلفا          42034.080   12138.69         -246.3
مساكن شيراتون     46575.100    7864.35         -492.2
عمرانيه 3         48615.170    5098.89         -853.4

[64 rows x 3 columns]

### part 5:

In [64]:
top = total_revenue.loc[:,'TotalRevenues'].sort_values(ascending = False )

In [65]:
top.to_frame().head()

TotalRevenues
district_ar                   
شبين القناطر              70.0
قلج                       35.6
بلقس                      33.5
الخانكة                   31.6
القناطر الخيرية           18.2

### part 6:

In [66]:
least = total_revenue.loc[:,'TotalRevenues'].sort_values(ascending = True )

In [67]:
least.to_frame().head()

TotalRevenues
district_ar                  
عمرانيه 3              -853.4
مساكن شيراتون          -492.2
ميت حلفا               -246.3
النزهة الجديدة         -145.2
الوايلي                -113.6

# Question 3: total monthly burn(discounts) for may and april

In [68]:
burn_april_may = df_AprilAndMay.groupby(['month'])['discount'].sum()

In [69]:
burn_april_may

month
4    194763.62
5    206311.08
Name: discount, dtype: float64

In [70]:
df_AprilAndMay.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  \
0  Application     العمرانية                      2119.75      0.00   
1  Application      ميت عقبة                      2242.75      0.10   
2  Application  مدينة السلام                      1094.50      8.65   
3  Application      ميت عقبة                      3448.50     24.51   
4  Application       المطرية                       910.95      0.10   

   order_price  
0      2119.75  
1      2242.65  
2      1085.85  
3      3423.99  
4       910.85

# Question 4: weekly contribution of every channel in the total weekly orders of may and april

In [71]:
WeeklyContriutionOfEachChannelInAprilAndMay_percentage = pd.pivot_table(df_AprilAndMay, values = 'sales_order_id', index = ['month', 'week'], columns= ['channel'], aggfunc = lambda x: round(((x.count()/df_AprilAndMay['sales_order_id'].count())*100),2))

In [72]:
WeeklyContriutionOfEachChannelInAprilAndMay_percentage

channel     Application  salesagent  telesales
month week                                    
4     14           3.29        0.31       0.42
      15           9.46        0.74       1.43
      16          11.27        0.84       1.34
      17          11.18        0.48       1.13
      18           7.73        0.87       0.70
5     18           0.98        0.02       0.09
      19           9.81        2.47       1.00
      20          10.01        2.20       1.06
      21          10.47        1.48       1.12
      22           4.26        0.35       0.28
      23           2.66        0.26       0.29

In [73]:
WeeklyContriutionOfEachChannelInAprilAndMay_count = pd.pivot_table(df_AprilAndMay, values = 'sales_order_id', index = ['month', 'week'], columns= ['channel'], aggfunc = 'count')

In [74]:
WeeklyContriutionOfEachChannelInAprilAndMay_count

channel     Application  salesagent  telesales
month week                                    
4     14            557          53         71
      15           1600         126        242
      16           1907         142        227
      17           1891          81        191
      18           1308         147        118
5     18            165           4         15
      19           1660         418        170
      20           1694         372        180
      21           1771         250        189
      22            720          59         47
      23            450          44         49

# Question 5:

In [75]:
df_AprilAndMay.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  \
0  Application     العمرانية                      2119.75      0.00   
1  Application      ميت عقبة                      2242.75      0.10   
2  Application  مدينة السلام                      1094.50      8.65   
3  Application      ميت عقبة                      3448.50     24.51   
4  Application       المطرية                       910.95      0.10   

   order_price  
0      2119.75  
1      2242.65  
2      1085.85  
3      3423.99  
4       910.85

In [76]:
df_AprilAndMay['sales_order_status'].value_counts()

Delivered    14207
Canceled      2127
Failed         584
Name: sales_order_status, dtype: int64

In [77]:
delivered_data = df_AprilAndMay[df_AprilAndMay['sales_order_status'] == 'Delivered'].reset_index(drop = True)

In [78]:
delivered_data

week  month creation_date  sales_order_id  retailer_id  \
0        14      4    2020-04-01               1        18564   
1        14      4    2020-04-01               5        28140   
2        14      4    2020-04-01              13        19299   
3        14      4    2020-04-01              17         5901   
4        14      4    2020-04-01              29        17783   
...     ...    ...           ...             ...          ...   
14202    23      5    2020-05-31           67664        17123   
14203    23      5    2020-05-31           67668        30589   
14204    23      5    2020-05-31           67672        24375   
14205    23      5    2020-05-31           67676        13651   
14206    23      5    2020-05-31           67680         5374   

      sales_order_status      channel    district_ar  \
0              Delivered  Application      العمرانية   
1              Delivered  Application       ميت عقبة   
2              Delivered  Application       ميت عقبة   
3              Delivered  Application        المطرية   
4              Delivered  Application          هرم 3   
...                  ...          ...            ...   
14202          Delivered    telesales          هرم 2   
14203          Delivered  Application           فيصل   
14204          Delivered  Application        عين شمس   
14205          Delivered  Application  بولاق الدكرور   
14206          Delivered  Application         امبابة   

       order_price_before_discount  discount  order_price  
0                          2119.75      0.00      2119.75  
1                          2242.75      0.10      2242.65  
2                          3448.50     24.51      3423.99  
3                           910.95      0.10       910.85  
4                          2113.45      0.10      2113.35  
...                            ...       ...          ...  
14202                      1922.25     12.82      1909.43  
14203                      1367.00      0.00      1367.00  
14204                      2180.50     39.09      2141.41  
14205                      3097.20     50.74      3046.46  
14206                      3746.70     33.89      3712.81  

[14207 rows x 11 columns]

In [79]:
delivered_data_may = delivered_data[delivered_data['month'] == 5].reset_index(drop = True)

In [80]:
delivered_data_may

week  month creation_date  sales_order_id  retailer_id  \
0       18      5    2020-05-01           34647        26658   
1       18      5    2020-05-01           34651         8913   
2       18      5    2020-05-01           34655        20805   
3       18      5    2020-05-01           34659        24977   
4       18      5    2020-05-01           34663         8544   
...    ...    ...           ...             ...          ...   
6757    23      5    2020-05-31           67664        17123   
6758    23      5    2020-05-31           67668        30589   
6759    23      5    2020-05-31           67672        24375   
6760    23      5    2020-05-31           67676        13651   
6761    23      5    2020-05-31           67680         5374   

     sales_order_status      channel    district_ar  \
0             Delivered  Application         المقطم   
1             Delivered  Application        المطرية   
2             Delivered  Application    باب الشعرية   
3             Delivered  Application        عين شمس   
4             Delivered  Application    حدائق القبة   
...                 ...          ...            ...   
6757          Delivered    telesales          هرم 2   
6758          Delivered  Application           فيصل   
6759          Delivered  Application        عين شمس   
6760          Delivered  Application  بولاق الدكرور   
6761          Delivered  Application         امبابة   

      order_price_before_discount  discount  order_price  
0                         1359.20      3.59      1355.61  
1                         3300.45     14.33      3286.12  
2                        10947.15     39.51     10907.64  
3                         3463.75     23.10      3440.65  
4                         2981.75     29.07      2952.68  
...                           ...       ...          ...  
6757                      1922.25     12.82      1909.43  
6758                      1367.00      0.00      1367.00  
6759                      2180.50     39.09      2141.41  
6760                      3097.20     50.74      3046.46  
6761                      3746.70     33.89      3712.81  

[6762 rows x 11 columns]

In [81]:
# two delivered app orders monthly

deliever_data_may = []
two_delivered_orders_may = []
for index, row in delivered_data_may.iterrows():
    if row['retailer_id'] not in deliever_data_may:
        deliever_data_may.append(row['retailer_id'])
    else:
        two_delivered_orders_may.append(row['retailer_id'])

delivered_data_may['two_delivered_orders_in_may'] = delivered_data_may['retailer_id'].isin(two_delivered_orders_may)

In [82]:
delivered_data_may

week  month creation_date  sales_order_id  retailer_id  \
0       18      5    2020-05-01           34647        26658   
1       18      5    2020-05-01           34651         8913   
2       18      5    2020-05-01           34655        20805   
3       18      5    2020-05-01           34659        24977   
4       18      5    2020-05-01           34663         8544   
...    ...    ...           ...             ...          ...   
6757    23      5    2020-05-31           67664        17123   
6758    23      5    2020-05-31           67668        30589   
6759    23      5    2020-05-31           67672        24375   
6760    23      5    2020-05-31           67676        13651   
6761    23      5    2020-05-31           67680         5374   

     sales_order_status      channel    district_ar  \
0             Delivered  Application         المقطم   
1             Delivered  Application        المطرية   
2             Delivered  Application    باب الشعرية   
3             Delivered  Application        عين شمس   
4             Delivered  Application    حدائق القبة   
...                 ...          ...            ...   
6757          Delivered    telesales          هرم 2   
6758          Delivered  Application           فيصل   
6759          Delivered  Application        عين شمس   
6760          Delivered  Application  بولاق الدكرور   
6761          Delivered  Application         امبابة   

      order_price_before_discount  discount  order_price  \
0                         1359.20      3.59      1355.61   
1                         3300.45     14.33      3286.12   
2                        10947.15     39.51     10907.64   
3                         3463.75     23.10      3440.65   
4                         2981.75     29.07      2952.68   
...                           ...       ...          ...   
6757                      1922.25     12.82      1909.43   
6758                      1367.00      0.00      1367.00   
6759                      2180.50     39.09      2141.41   
6760                      3097.20     50.74      3046.46   
6761                      3746.70     33.89      3712.81   

      two_delivered_orders_in_may  
0                            True  
1                           False  
2                            True  
3                           False  
4                           False  
...                           ...  
6757                        False  
6758                         True  
6759                         True  
6760                         True  
6761                         True  

[6762 rows x 12 columns]

In [83]:
delivered_data_may['two_delivered_orders_in_may'].value_counts()

True     3522
False    3240
Name: two_delivered_orders_in_may, dtype: int64

### percentage of application users

In [84]:
delieverd_app_may_data = df_AprilAndMay[(df_AprilAndMay['month'] == 5) & (df_AprilAndMay['channel'] == 'Application') & (df_AprilAndMay['sales_order_status'] == 'Delivered')]

In [85]:
delieverd_app_may_data

week  month creation_date  sales_order_id  retailer_id  \
8661     18      5    2020-05-01           34647        26658   
8662     18      5    2020-05-01           34651         8913   
8663     18      5    2020-05-01           34655        20805   
8664     18      5    2020-05-01           34659        24977   
8665     18      5    2020-05-01           34663         8544   
...     ...    ...           ...             ...          ...   
16912    23      5    2020-05-31           67660        18413   
16914    23      5    2020-05-31           67668        30589   
16915    23      5    2020-05-31           67672        24375   
16916    23      5    2020-05-31           67676        13651   
16917    23      5    2020-05-31           67680         5374   

      sales_order_status      channel    district_ar  \
8661           Delivered  Application         المقطم   
8662           Delivered  Application        المطرية   
8663           Delivered  Application    باب الشعرية   
8664           Delivered  Application        عين شمس   
8665           Delivered  Application    حدائق القبة   
...                  ...          ...            ...   
16912          Delivered  Application         فيصل 3   
16914          Delivered  Application           فيصل   
16915          Delivered  Application        عين شمس   
16916          Delivered  Application  بولاق الدكرور   
16917          Delivered  Application         امبابة   

       order_price_before_discount  discount  order_price  
8661                       1359.20      3.59      1355.61  
8662                       3300.45     14.33      3286.12  
8663                      10947.15     39.51     10907.64  
8664                       3463.75     23.10      3440.65  
8665                       2981.75     29.07      2952.68  
...                            ...       ...          ...  
16912                      2541.75     42.95      2498.80  
16914                      1367.00      0.00      1367.00  
16915                      2180.50     39.09      2141.41  
16916                      3097.20     50.74      3046.46  
16917                      3746.70     33.89      3712.81  

[5215 rows x 11 columns]

In [86]:
delieverd_app_may_data.count()[1]

5215

In [87]:
delievered_app_may_orders_counts = pd.pivot_table(delieverd_app_may_data, values = 'sales_order_id', index = 'retailer_id', columns = 'month', aggfunc = 'count')
delievered_app_may_orders_counts

month        5
retailer_id   
96           1
126          1
130          1
136          1
242          1
...         ..
38710        1
38711        1
38737        1
38768        1
38941        1

[3372 rows x 1 columns]

In [88]:
delievered_app_may_orders_counts[5].value_counts()

1    2210
2     735
3     264
4      96
5      49
6      12
7       6
Name: 5, dtype: int64

In [89]:
delievered_may_orders_counts = pd.pivot_table(delivered_data_may, values = 'sales_order_id', index = 'retailer_id', columns = 'month', aggfunc = 'count')
delievered_may_orders_counts

month        5
retailer_id   
85           1
96           1
126          1
129          2
130          1
...         ..
38946        1
38952        1
38983        1
38986        1
39003        1

[4621 rows x 1 columns]

In [90]:
delievered_may_orders_counts[5].value_counts()

1    3240
2     899
3     302
4     106
5      56
6      12
7       6
Name: 5, dtype: int64

In [91]:
delievered_app_may_orders_percentage_from_total = np.round((delievered_app_may_orders_counts / delievered_may_orders_counts)*100,1)
delievered_app_may_orders_percentage_from_total.fillna(0, inplace = True)

In [92]:
delievered_app_may_orders_percentage_from_total

month            5
retailer_id       
85             0.0
96           100.0
126          100.0
129            0.0
130          100.0
...            ...
38946          0.0
38952          0.0
38983          0.0
38986          0.0
39003          0.0

[4621 rows x 1 columns]

In [93]:
delievered_app_may_orders_percentage_from_total[5].value_counts()

100.0    3277
0.0      1249
50.0       71
66.7        9
33.3        7
75.0        3
80.0        2
25.0        1
60.0        1
20.0        1
Name: 5, dtype: int64

In [94]:
above_75 = []
for index, row in delievered_app_may_orders_percentage_from_total.iterrows():
    if row[5] > 75.0:
        above_75.append(index)
    else:
        continue
delivered_data_may['above_75'] = delivered_data_may['retailer_id'].isin(above_75)

In [95]:
delivered_data_may

week  month creation_date  sales_order_id  retailer_id  \
0       18      5    2020-05-01           34647        26658   
1       18      5    2020-05-01           34651         8913   
2       18      5    2020-05-01           34655        20805   
3       18      5    2020-05-01           34659        24977   
4       18      5    2020-05-01           34663         8544   
...    ...    ...           ...             ...          ...   
6757    23      5    2020-05-31           67664        17123   
6758    23      5    2020-05-31           67668        30589   
6759    23      5    2020-05-31           67672        24375   
6760    23      5    2020-05-31           67676        13651   
6761    23      5    2020-05-31           67680         5374   

     sales_order_status      channel    district_ar  \
0             Delivered  Application         المقطم   
1             Delivered  Application        المطرية   
2             Delivered  Application    باب الشعرية   
3             Delivered  Application        عين شمس   
4             Delivered  Application    حدائق القبة   
...                 ...          ...            ...   
6757          Delivered    telesales          هرم 2   
6758          Delivered  Application           فيصل   
6759          Delivered  Application        عين شمس   
6760          Delivered  Application  بولاق الدكرور   
6761          Delivered  Application         امبابة   

      order_price_before_discount  discount  order_price  \
0                         1359.20      3.59      1355.61   
1                         3300.45     14.33      3286.12   
2                        10947.15     39.51     10907.64   
3                         3463.75     23.10      3440.65   
4                         2981.75     29.07      2952.68   
...                           ...       ...          ...   
6757                      1922.25     12.82      1909.43   
6758                      1367.00      0.00      1367.00   
6759                      2180.50     39.09      2141.41   
6760                      3097.20     50.74      3046.46   
6761                      3746.70     33.89      3712.81   

      two_delivered_orders_in_may  above_75  
0                            True      True  
1                           False      True  
2                            True      True  
3                           False      True  
4                           False      True  
...                           ...       ...  
6757                        False     False  
6758                         True      True  
6759                         True      True  
6760                         True      True  
6761                         True      True  

[6762 rows x 13 columns]

In [96]:
may_app_users = delivered_data_may[(delivered_data_may['two_delivered_orders_in_may'] == True) & (delivered_data_may['above_75'] == True)].reset_index(drop = True)

In [97]:
may_app_users

week  month creation_date  sales_order_id  retailer_id  \
0       18      5    2020-05-01           34647        26658   
1       18      5    2020-05-01           34655        20805   
2       18      5    2020-05-01           34667        17216   
3       18      5    2020-05-01           34671        13813   
4       18      5    2020-05-01           34683        35014   
...    ...    ...           ...             ...          ...   
2970    23      5    2020-05-31           67652        33954   
2971    23      5    2020-05-31           67668        30589   
2972    23      5    2020-05-31           67672        24375   
2973    23      5    2020-05-31           67676        13651   
2974    23      5    2020-05-31           67680         5374   

     sales_order_status      channel    district_ar  \
0             Delivered  Application         المقطم   
1             Delivered  Application    باب الشعرية   
2             Delivered  Application      العمرانية   
3             Delivered  Application  بولاق الدكرور   
4             Delivered  Application    باب الشعرية   
...                 ...          ...            ...   
2970          Delivered  Application         امبابة   
2971          Delivered  Application           فيصل   
2972          Delivered  Application        عين شمس   
2973          Delivered  Application  بولاق الدكرور   
2974          Delivered  Application         امبابة   

      order_price_before_discount  discount  order_price  \
0                         1359.20      3.59      1355.61   
1                        10947.15     39.51     10907.64   
2                          928.00      0.00       928.00   
3                        15427.25    294.85     15132.40   
4                         1412.00      0.00      1412.00   
...                           ...       ...          ...   
2970                      1246.75      0.00      1246.75   
2971                      1367.00      0.00      1367.00   
2972                      2180.50     39.09      2141.41   
2973                      3097.20     50.74      3046.46   
2974                      3746.70     33.89      3712.81   

      two_delivered_orders_in_may  above_75  
0                            True      True  
1                            True      True  
2                            True      True  
3                            True      True  
4                            True      True  
...                           ...       ...  
2970                         True      True  
2971                         True      True  
2972                         True      True  
2973                         True      True  
2974                         True      True  

[2975 rows x 13 columns]

In [98]:
round((may_app_users.count()[1] / delivered_data_may.count()[1])*100,0)

44.0

so: % of app users = 44%

### percentage of telesales users 

In [99]:
len(delivered_data['retailer_id'].unique())

7080

In [100]:
last_channel_record = delivered_data.groupby(['retailer_id'])['channel'].last().to_frame()

In [101]:
last_channel_record

channel
retailer_id             
68           Application
73           Application
80           Application
85             telesales
90             telesales
...                  ...
38946         salesagent
38952         salesagent
38983         salesagent
38986         salesagent
39003         salesagent

[7080 rows x 1 columns]

In [102]:
telesales_last_record = last_channel_record[last_channel_record['channel'] == 'telesales']

In [103]:
telesales_last_record

channel
retailer_id           
85           telesales
90           telesales
109          telesales
129          telesales
151          telesales
...                ...
37742        telesales
38100        telesales
38119        telesales
38275        telesales
38657        telesales

[761 rows x 1 columns]

In [104]:
tele_retailer_list = []
for index, row in telesales_last_record.iterrows():
    if index not in tele_retailer_list:
        tele_retailer_list.append(index)
    else:
        continue

In [105]:
delivered_data_may['is_last_tele'] = delivered_data_may['retailer_id'].isin(tele_retailer_list)

In [106]:
delivered_data_may

week  month creation_date  sales_order_id  retailer_id  \
0       18      5    2020-05-01           34647        26658   
1       18      5    2020-05-01           34651         8913   
2       18      5    2020-05-01           34655        20805   
3       18      5    2020-05-01           34659        24977   
4       18      5    2020-05-01           34663         8544   
...    ...    ...           ...             ...          ...   
6757    23      5    2020-05-31           67664        17123   
6758    23      5    2020-05-31           67668        30589   
6759    23      5    2020-05-31           67672        24375   
6760    23      5    2020-05-31           67676        13651   
6761    23      5    2020-05-31           67680         5374   

     sales_order_status      channel    district_ar  \
0             Delivered  Application         المقطم   
1             Delivered  Application        المطرية   
2             Delivered  Application    باب الشعرية   
3             Delivered  Application        عين شمس   
4             Delivered  Application    حدائق القبة   
...                 ...          ...            ...   
6757          Delivered    telesales          هرم 2   
6758          Delivered  Application           فيصل   
6759          Delivered  Application        عين شمس   
6760          Delivered  Application  بولاق الدكرور   
6761          Delivered  Application         امبابة   

      order_price_before_discount  discount  order_price  \
0                         1359.20      3.59      1355.61   
1                         3300.45     14.33      3286.12   
2                        10947.15     39.51     10907.64   
3                         3463.75     23.10      3440.65   
4                         2981.75     29.07      2952.68   
...                           ...       ...          ...   
6757                      1922.25     12.82      1909.43   
6758                      1367.00      0.00      1367.00   
6759                      2180.50     39.09      2141.41   
6760                      3097.20     50.74      3046.46   
6761                      3746.70     33.89      3712.81   

      two_delivered_orders_in_may  above_75  is_last_tele  
0                            True      True         False  
1                           False      True         False  
2                            True      True         False  
3                           False      True         False  
4                           False      True         False  
...                           ...       ...           ...  
6757                        False     False          True  
6758                         True      True         False  
6759                         True      True         False  
6760                         True      True         False  
6761                         True      True         False  

[6762 rows x 14 columns]

In [107]:
may_telesales_users = delivered_data_may[(delivered_data_may['two_delivered_orders_in_may'] == True) & (delivered_data_may['is_last_tele'] == True)].reset_index(drop = True)

In [108]:
may_telesales_users

week  month creation_date  sales_order_id  retailer_id  \
0      18      5    2020-05-01           34815         6961   
1      18      5    2020-05-01           35019        15611   
2      18      5    2020-05-01           35023         6262   
3      18      5    2020-05-01           35039        26445   
4      18      5    2020-05-01           35059        12129   
..    ...    ...           ...             ...          ...   
300    23      5    2020-05-31           67052        35651   
301    23      5    2020-05-31           67104        20400   
302    23      5    2020-05-31           67220         9826   
303    23      5    2020-05-31           67288         8063   
304    23      5    2020-05-31           67500        22280   

    sales_order_status    channel      district_ar  \
0            Delivered  telesales           امبابة   
1            Delivered  telesales             فيصل   
2            Delivered  telesales           الوراق   
3            Delivered  telesales          السبتية   
4            Delivered  telesales          عين شمس   
..                 ...        ...              ...   
300          Delivered  telesales             فيصل   
301          Delivered  telesales         العباسية   
302          Delivered  telesales  الزاوية الحمراء   
303          Delivered  telesales      شبرا الخيمة   
304          Delivered  telesales            المرج   

     order_price_before_discount  discount  order_price  \
0                        2013.00     27.78      1985.22   
1                        2951.30     16.43      2934.87   
2                        8833.75    114.30      8719.45   
3                        2811.50     19.45      2792.05   
4                        1163.05      9.15      1153.90   
..                           ...       ...          ...   
300                      4613.65     47.57      4566.08   
301                      1149.00     15.03      1133.97   
302                      2472.50     29.69      2442.81   
303                       922.20      8.82       913.38   
304                      1915.50     16.54      1898.96   

     two_delivered_orders_in_may  above_75  is_last_tele  
0                           True     False          True  
1                           True     False          True  
2                           True     False          True  
3                           True     False          True  
4                           True     False          True  
..                           ...       ...           ...  
300                         True     False          True  
301                         True     False          True  
302                         True     False          True  
303                         True     False          True  
304                         True     False          True  

[305 rows x 14 columns]

In [109]:
round((may_telesales_users.count()[1] / delivered_data_may.count()[1])*100,0)

5.0

so: % of telesales users = 5%

# Question 6:

In [110]:
df_copy.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  
0  Application     العمرانية                      2119.75      0.00  
1  Application      ميت عقبة                      2242.75      0.10  
2  Application  مدينة السلام                      1094.50      8.65  
3  Application      ميت عقبة                      3448.50     24.51  
4  Application       المطرية                       910.95      0.10

In [111]:
df_copy['revenue'] = df_copy['order_price_before_discount'] - df_copy['discount']

In [112]:
df_copy.head()

week  month creation_date  sales_order_id  retailer_id sales_order_status  \
0    14      4    2020-04-01               1        18564          Delivered   
1    14      4    2020-04-01               5        28140          Delivered   
2    14      4    2020-04-01               9        32552           Canceled   
3    14      4    2020-04-01              13        19299          Delivered   
4    14      4    2020-04-01              17         5901          Delivered   

       channel   district_ar  order_price_before_discount  discount  revenue  
0  Application     العمرانية                      2119.75      0.00  2119.75  
1  Application      ميت عقبة                      2242.75      0.10  2242.65  
2  Application  مدينة السلام                      1094.50      8.65  1085.85  
3  Application      ميت عقبة                      3448.50     24.51  3423.99  
4  Application       المطرية                       910.95      0.10   910.85

In [161]:
df_copy.tail()

week  month creation_date  sales_order_id  retailer_id  \
21711    25      6    2020-06-15           86872        30883   
21712    25      6    2020-06-15           86876         5619   
21713    25      6    2020-06-15           86880        16638   
21714    25      6    2020-06-15           86884        36652   
21715    25      6    2020-06-15           86888        39303   

      sales_order_status      channel district_ar  \
21711          Delivered  Application       الدقي   
21712           Canceled  Application      الخصوص   
21713          Delivered  Application        فيصل   
21714          Delivered  Application       مسطرد   
21715          Delivered  Application  دار السلام   

       order_price_before_discount  discount  revenue  
21711                      3309.00     46.52  3262.48  
21712                      2805.10     43.69  2761.41  
21713                      2294.20     39.76  2254.44  
21714                      1660.35     18.92  1641.43  
21715                       966.60      0.10   966.50

In [168]:
retailers_sales_orders = pd.pivot_table(df_copy, index = 'retailer_id', values ='revenue', columns = ['month'], aggfunc = 'sum')

In [140]:
retailers_sales_orders.std()

month  week
4      14      3209.872211
       15      4002.324570
       16      3981.367358
       17      3637.230287
       18      2726.348465
5      18      2216.710371
       19      3513.430021
       20      2736.241668
       21      3996.599963
       22      3269.366053
       23      2532.399588
6      23      3210.503904
       24      3491.569232
       25      2191.403794
dtype: float64

In [141]:
retailers_sales_orders.mean()

month  week
4      14      3381.876296
       15      3760.651019
       16      3486.534182
       17      3410.677596
       18      2738.596482
5      18      2713.174918
       19      2661.141418
       20      2601.848871
       21      3425.689972
       22      3476.054854
       23      2707.893404
6      23      2695.785918
       24      2612.781544
       25      2429.112383
dtype: float64

In [142]:
retailers_sales_orders.max()

month  week
4      14      27281.91
       15      59262.77
       16      46894.91
       17      64513.06
       18      32362.10
5      18      15132.40
       19      71140.90
       20      45362.61
       21      52999.17
       22      32321.66
       23      24701.99
6      23      62737.17
       24      98316.00
       25      26824.53
dtype: float64

In [143]:
retailers_sales_orders.min()

month  week
4      14         0.0
       15       150.0
       16     -3153.4
       17         0.0
       18         0.0
5      18         0.0
       19        -0.1
       20       272.4
       21        -0.1
       22       235.0
       23       251.0
6      23       301.0
       24        -0.1
       25       321.0
dtype: float64

In [164]:
retailers_sales_orders.unstack()
pd.set_option('display.max_rows',10)

In [165]:
retailers_sales_orders

month             4                                5                           \
week             14  15     16       17       18  18  19       20  21  22  23   
retailer_id                                                                     
68              NaN NaN    NaN  1051.28      NaN NaN NaN      NaN NaN NaN NaN   
73           1770.9 NaN    NaN      NaN      NaN NaN NaN      NaN NaN NaN NaN   
80              NaN NaN  843.3      NaN      NaN NaN NaN      NaN NaN NaN NaN   
85              NaN NaN    NaN  6700.45      NaN NaN NaN  9818.81 NaN NaN NaN   
90              NaN NaN    NaN      NaN  1366.38 NaN NaN      NaN NaN NaN NaN   
...             ...  ..    ...      ...      ...  ..  ..      ...  ..  ..  ..   
40745           NaN NaN    NaN      NaN      NaN NaN NaN      NaN NaN NaN NaN   
40753           NaN NaN    NaN      NaN      NaN NaN NaN      NaN NaN NaN NaN   
40758           NaN NaN    NaN      NaN      NaN NaN NaN      NaN NaN NaN NaN   
40770           NaN NaN    NaN      NaN      NaN NaN NaN      NaN NaN NaN NaN   
40778           NaN NaN    NaN      NaN      NaN NaN NaN      NaN NaN NaN NaN   

month         6               
week         23  24       25  
retailer_id                   
68          NaN NaN      NaN  
73          NaN NaN      NaN  
80          NaN NaN      NaN  
85          NaN NaN      NaN  
90          NaN NaN      NaN  
...          ..  ..      ...  
40745       NaN NaN   823.25  
40753       NaN NaN   847.70  
40758       NaN NaN   933.30  
40770       NaN NaN   932.00  
40778       NaN NaN  1661.56  

[9276 rows x 14 columns]

In [166]:
retailers_sales_orders.unstack()

month  week  retailer_id
4      14    68                 NaN
             73             1770.90
             80                 NaN
             85                 NaN
             90                 NaN
                             ...   
6      25    40745           823.25
             40753           847.70
             40758           933.30
             40770           932.00
             40778          1661.56
Length: 129864, dtype: float64

In [175]:
retailers_sales_orders.fillna(0, inplace = True)

In [176]:
# if there are sales in may and april so may we will have sales on june
# if there are sales on may only so may we will have sales on june too
# if there are sales on april so may we will have sales on june too
# if there are no sales on both may and april so may it means that the retailer won't have sales on june

june_retailers_sales_withNoMayOrAprilSales = retailers_sales_orders[(retailers_sales_orders[6] != 0) | ((retailers_sales_orders[5] == 0) & (retailers_sales_orders[4] == 0))]

In [177]:
june_retailers_sales_withNoMayOrAprilSales

month              4        5         6
retailer_id                            
98           2471.00     0.00   3618.69
120          3096.69     0.00   1182.26
129             0.00  9257.69  12886.40
209             0.00     0.00   3013.76
217             0.00     0.00    830.75
...              ...      ...       ...
40745           0.00     0.00    823.25
40753           0.00     0.00    847.70
40758           0.00     0.00    933.30
40770           0.00     0.00    932.00
40778           0.00     0.00   1661.56

[3681 rows x 3 columns]

In [178]:
# extract actual retailers from data and try to estimate our potential retailers june sales
june_empty_sales = retailers_sales_orders[retailers_sales_orders[6] == 0]


In [179]:
june_empty_sales

month              4        5    6
retailer_id                       
68           1051.28     0.00  0.0
73           1770.90     0.00  0.0
80            843.30     0.00  0.0
85           6700.45  9818.81  0.0
90           1366.38     0.00  0.0
...              ...      ...  ...
38946           0.00  2313.07  0.0
38952           0.00  1537.50  0.0
38983           0.00  2185.71  0.0
38986           0.00  6125.98  0.0
39003           0.00  2384.96  0.0

[5601 rows x 3 columns]

In [182]:
# expecting june sales for those retailers using average sales in both may and april

just_june_empty = june_empty_sales[6].to_frame()

In [185]:
just_june_fill = pd.DataFrame(index = just_june_empty.index, columns = [6])

In [187]:
just_june_fill[6] = np.round((june_empty_sales[4] + june_empty_sales[5]) / 2)

In [188]:
just_june_fill

6
retailer_id        
68            526.0
73            885.0
80            422.0
85           8260.0
90            683.0
...             ...
38946        1157.0
38952         769.0
38983        1093.0
38986        3063.0
39003        1192.0

[5601 rows x 1 columns]

In [189]:
just_june_retailers_sales_withNoMayOrAprilSales = june_retailers_sales_withNoMayOrAprilSales[6].to_frame()

In [213]:
all_june_retailers_sales = pd.concat([just_june_retailers_sales_withNoMayOrAprilSales, just_june_fill], axis = 0).reset_index()

In [214]:
all_june_retailers_sales

retailer_id         6
0              98   3618.69
1             120   1182.26
2             129  12886.40
3             209   3013.76
4             217    830.75
...           ...       ...
9277        38946   1157.00
9278        38952    769.00
9279        38983   1093.00
9280        38986   3063.00
9281        39003   1192.00

[9282 rows x 2 columns]

In [218]:
all_june_sales = all_june_retailers_sales.drop_duplicates(subset = 'retailer_id', keep = 'first')

In [225]:
all_june_sales.sort_values('retailer_id', ascending = True, inplace = True)

In [228]:
all_june_sales.reset_index(drop = True, inplace = True)

In [230]:
all_june_sales

retailer_id        6
0              68   526.00
1              73   885.00
2              80   422.00
3              85  8260.00
4              90   683.00
...           ...      ...
9271        40745   823.25
9272        40753   847.70
9273        40758   933.30
9274        40770   932.00
9275        40778  1661.56

[9276 rows x 2 columns]

In [233]:
all_june_sales[all_june_sales['retailer_id'] == 16861]

retailer_id    6
3092        16861  0.0

In [235]:
all_june_sales[all_june_sales['retailer_id'] == 38946]

retailer_id       6
8927        38946  1157.0

In [236]:
all_june_sales[all_june_sales['retailer_id'] == 7819]

retailer_id         6
956         7819  62737.17

# Question 7:

In [196]:
df2_stock = pd.read_excel(r'C:\Users\ycc\Downloads\MaxAB Task\dataset2.xlsx', sheet_name='Stocks')
df2_sales = pd.read_excel(r'C:\Users\ycc\Downloads\MaxAB Task\dataset2.xlsx', sheet_name='Sales')

In [197]:
df2_stock

sku  Stock
0                          أرز حبوبة رفيع - 1 كيلو   1395
1            زيت كريستال ذرة ( 6 أزازة ) - 2.4 لتر     58
2     زيت كريستال عباد الشمس ( 6 أزازة ) - 2.4 لتر     43
3                 زيت كريستال الممتاز خليط - 1 لتر     47
4                             سكر الأهرام - 1 كيلو   1757
...                                            ...    ...
6541    شاى ليبتون أخضر بدون مرارة (عرض) - 25 فتلة      3
6542                      شويبس رمان كانز - 300 مل    505
6543               شويبس جولد أناناس كانز - 300 مل    323
6544              حفاضات مولفيكس (58) عرض - 4 مقاس     10
6545              حفاضات مولفيكس (58) عرض - 3 مقاس     81

[6546 rows x 2 columns]

In [198]:
df2_sales

Date  Week                                           sku  \
0    2020-06-06    24                       أرز حبوبة رفيع - 1 كيلو   
1    2020-06-06    24         زيت كريستال ذرة ( 6 أزازة ) - 2.4 لتر   
2    2020-06-06    24  زيت كريستال عباد الشمس ( 6 أزازة ) - 2.4 لتر   
3    2020-06-06    24              زيت كريستال الممتاز خليط - 1 لتر   
4    2020-06-06    24                          سكر الأهرام - 1 كيلو   
...         ...   ...                                           ...   
6541 2020-06-15    25    شاى ليبتون أخضر بدون مرارة (عرض) - 25 فتلة   
6542 2020-06-15    25                      شويبس رمان كانز - 300 مل   
6543 2020-06-15    25               شويبس جولد أناناس كانز - 300 مل   
6544 2020-06-15    25              حفاضات مولفيكس (58) عرض - 4 مقاس   
6545 2020-06-15    25              حفاضات مولفيكس (58) عرض - 3 مقاس   

      Sales_Count  
0              69  
1              17  
2              14  
3               7  
4              78  
...           ...  
6541           74  
6542           79  
6543           44  
6544            2  
6545            1  

[6546 rows x 4 columns]

In [200]:
sales_per_sku = pd.pivot_table(df2_sales, values = 'Sales_Count', index = 'sku', aggfunc = 'sum' )

In [201]:
sales_per_sku

Sales_Count
sku                                                           
AA سوني زنك (شرنك) - 4 حجر                                   8
AA فيليبس (شرنك) - 4 حجر                                    20
AAAفيليبس الكالين كارت - 2 حجر                               1
D فيليبس (شرنك) - 2 حجر                                      9
أرز  الضحى مصرى - 5 كيلو                                    46
...                                                        ...
ويفر فانيليا بسكو مصر - 2 جنية                             186
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية                     221
ويفر كتاكيتو ميني - 3 جنية                                 207
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية            1
ويفرالشمعدان اخضر فاخر - 2 جنية                             91

[897 rows x 1 columns]

In [ ]:
sales_per_sku['Stock'] = df2_stock['']

In [212]:
df2_stock.groupby(['sku'])['Stock'].sum().to_frame()

Stock
sku                                                     
AA سوني زنك (شرنك) - 4 حجر                           234
AA فيليبس (شرنك) - 4 حجر                             300
AAAفيليبس الكالين كارت - 2 حجر                        15
D فيليبس (شرنك) - 2 حجر                              348
أرز  الضحى مصرى - 5 كيلو                            1596
...                                                  ...
ويفر فانيليا بسكو مصر - 2 جنية                      4301
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية              6432
ويفر كتاكيتو ميني - 3 جنية                          8216
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية     26
ويفرالشمعدان اخضر فاخر - 2 جنية                     2210

[897 rows x 1 columns]

In [213]:
df2_stock.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
6541     True
6542     True
6543     True
6544     True
6545    False
Length: 6546, dtype: bool

In [215]:
df2_stock.drop_duplicates(inplace = True)

In [216]:
df2_stock

sku  Stock
0                          أرز حبوبة رفيع - 1 كيلو   1395
1            زيت كريستال ذرة ( 6 أزازة ) - 2.4 لتر     58
2     زيت كريستال عباد الشمس ( 6 أزازة ) - 2.4 لتر     43
3                 زيت كريستال الممتاز خليط - 1 لتر     47
4                             سكر الأهرام - 1 كيلو   1757
...                                            ...    ...
6395                   شربات ورد فيتراك - 845 جرام     13
6492                  فول تدميس حبة حبة - 500 جرام     53
6536          مسحوق برسيل أوتوماتيك عرض - 2.5 كيلو     10
6538         مناديل بكر تواليت دولفين فاخر - 2 رول      3
6545              حفاضات مولفيكس (58) عرض - 3 مقاس     81

[897 rows x 2 columns]

In [224]:
sku_in_stock = df2_stock.groupby(['sku'])['Stock'].sum().to_frame()

In [225]:
sku_in_stock

Stock
sku                                                     
AA سوني زنك (شرنك) - 4 حجر                            78
AA فيليبس (شرنك) - 4 حجر                              50
AAAفيليبس الكالين كارت - 2 حجر                        15
D فيليبس (شرنك) - 2 حجر                               87
أرز  الضحى مصرى - 5 كيلو                             114
...                                                  ...
ويفر فانيليا بسكو مصر - 2 جنية                       253
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية               536
ويفر كتاكيتو ميني - 3 جنية                           632
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية     26
ويفرالشمعدان اخضر فاخر - 2 جنية                      221

[897 rows x 1 columns]

In [218]:
sales_per_sku

Sales_Count
sku                                                           
AA سوني زنك (شرنك) - 4 حجر                                   8
AA فيليبس (شرنك) - 4 حجر                                    20
AAAفيليبس الكالين كارت - 2 حجر                               1
D فيليبس (شرنك) - 2 حجر                                      9
أرز  الضحى مصرى - 5 كيلو                                    46
...                                                        ...
ويفر فانيليا بسكو مصر - 2 جنية                             186
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية                     221
ويفر كتاكيتو ميني - 3 جنية                                 207
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية            1
ويفرالشمعدان اخضر فاخر - 2 جنية                             91

[897 rows x 1 columns]

In [222]:
all_sales_data = pd.DataFrame(index = sales_per_sku.index, columns = ['Sales_Count', 'Stock', 'days_on_hands_per_sku'] )

In [223]:
all_sales_data

Sales_Count Stock  \
sku                                                                   
AA سوني زنك (شرنك) - 4 حجر                                NaN   NaN   
AA فيليبس (شرنك) - 4 حجر                                  NaN   NaN   
AAAفيليبس الكالين كارت - 2 حجر                            NaN   NaN   
D فيليبس (شرنك) - 2 حجر                                   NaN   NaN   
أرز  الضحى مصرى - 5 كيلو                                  NaN   NaN   
...                                                       ...   ...   
ويفر فانيليا بسكو مصر - 2 جنية                            NaN   NaN   
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية                    NaN   NaN   
ويفر كتاكيتو ميني - 3 جنية                                NaN   NaN   
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية         NaN   NaN   
ويفرالشمعدان اخضر فاخر - 2 جنية                           NaN   NaN   

                                                  days_on_hands_per_sku  
sku                                                                      
AA سوني زنك (شرنك) - 4 حجر                                          NaN  
AA فيليبس (شرنك) - 4 حجر                                            NaN  
AAAفيليبس الكالين كارت - 2 حجر                                      NaN  
D فيليبس (شرنك) - 2 حجر                                             NaN  
أرز  الضحى مصرى - 5 كيلو                                            NaN  
...                                                                 ...  
ويفر فانيليا بسكو مصر - 2 جنية                                      NaN  
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية                              NaN  
ويفر كتاكيتو ميني - 3 جنية                                          NaN  
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية                   NaN  
ويفرالشمعدان اخضر فاخر - 2 جنية                                     NaN  

[897 rows x 3 columns]

In [229]:
all_sales_data['Sales_Count'] = sales_per_sku['Sales_Count']
all_sales_data['Stock'] = sku_in_stock['Stock']
all_sales_data['days_on_hands_per_sku'] = np.round(all_sales_data['Stock'] / (all_sales_data['Sales_Count'] / 10))

In [230]:
all_sales_data

Sales_Count  Stock  \
sku                                                                     
AA سوني زنك (شرنك) - 4 حجر                                   8     78   
AA فيليبس (شرنك) - 4 حجر                                    20     50   
AAAفيليبس الكالين كارت - 2 حجر                               1     15   
D فيليبس (شرنك) - 2 حجر                                      9     87   
أرز  الضحى مصرى - 5 كيلو                                    46    114   
...                                                        ...    ...   
ويفر فانيليا بسكو مصر - 2 جنية                             186    253   
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية                     221    536   
ويفر كتاكيتو ميني - 3 جنية                                 207    632   
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية            1     26   
ويفرالشمعدان اخضر فاخر - 2 جنية                             91    221   

                                                   days_on_hands_per_sku  
sku                                                                       
AA سوني زنك (شرنك) - 4 حجر                                          98.0  
AA فيليبس (شرنك) - 4 حجر                                            25.0  
AAAفيليبس الكالين كارت - 2 حجر                                     150.0  
D فيليبس (شرنك) - 2 حجر                                             97.0  
أرز  الضحى مصرى - 5 كيلو                                            25.0  
...                                                                  ...  
ويفر فانيليا بسكو مصر - 2 جنية                                      14.0  
ويفر كتاكيتو مغطي بالشوكولاته - 5 جنية                              24.0  
ويفر كتاكيتو ميني - 3 جنية                                          31.0  
ويفر كوماندا بيور محشو بكريمه الشوكولاته - 2 جنية                  260.0  
ويفرالشمعدان اخضر فاخر - 2 جنية                                     24.0  

[897 rows x 3 columns]

# Question 8:

In [232]:
# select products with less than 7 days on hands per sku

sku_less_than_7days = all_sales_data[all_sales_data['days_on_hands_per_sku'] <= 7.0]

In [233]:
sku_less_than_7days

Sales_Count  Stock  \
sku                                                            
أرز الضحى مصري - 1 كيلو                           153     84   
أرز رفيع الأهرام - 1 كيلو                          50     25   
اريال جل - 2.5 لتر                                 18      2   
افيريدي ازرق AA - 20 حجر                           37     16   
افيريدي ازرق AAA - 40 حجر                          35     16   
...                                               ...    ...   
نسكافية بندق 3*1 ظرف (عرض) - 18 جرام               97      9   
نسكافية شوكولاتة 3*1 ظرف (عرض) - 18 جرام           99     52   
نسكافية ظرف 3*1 (عرض) - 18 جرام                    90     43   
هوت صوص هاينز - 165 مل                             81     39   
هيد أند شولدرز اكياس 2+1 (عرض) - 5 مل              39      5   

                                          days_on_hands_per_sku  
sku                                                              
أرز الضحى مصري - 1 كيلو                                     5.0  
أرز رفيع الأهرام - 1 كيلو                                   5.0  
اريال جل - 2.5 لتر                                          1.0  
افيريدي ازرق AA - 20 حجر                                    4.0  
افيريدي ازرق AAA - 40 حجر                                   5.0  
...                                                         ...  
نسكافية بندق 3*1 ظرف (عرض) - 18 جرام                        1.0  
نسكافية شوكولاتة 3*1 ظرف (عرض) - 18 جرام                    5.0  
نسكافية ظرف 3*1 (عرض) - 18 جرام                             5.0  
هوت صوص هاينز - 165 مل                                      5.0  
هيد أند شولدرز اكياس 2+1 (عرض) - 5 مل                       1.0  

[185 rows x 3 columns]

In [234]:
all_sales_data_SkuFor7Days = pd.DataFrame(index = sku_less_than_7days.index, columns = ['Sales_Count', 'Stock', 'desired_stock', 'to_be_purchased'] )

In [235]:
all_sales_data_SkuFor7Days

Sales_Count Stock desired_stock  \
sku                                                                        
أرز الضحى مصري - 1 كيلو                          NaN   NaN           NaN   
أرز رفيع الأهرام - 1 كيلو                        NaN   NaN           NaN   
اريال جل - 2.5 لتر                               NaN   NaN           NaN   
افيريدي ازرق AA - 20 حجر                         NaN   NaN           NaN   
افيريدي ازرق AAA - 40 حجر                        NaN   NaN           NaN   
...                                              ...   ...           ...   
نسكافية بندق 3*1 ظرف (عرض) - 18 جرام             NaN   NaN           NaN   
نسكافية شوكولاتة 3*1 ظرف (عرض) - 18 جرام         NaN   NaN           NaN   
نسكافية ظرف 3*1 (عرض) - 18 جرام                  NaN   NaN           NaN   
هوت صوص هاينز - 165 مل                           NaN   NaN           NaN   
هيد أند شولدرز اكياس 2+1 (عرض) - 5 مل            NaN   NaN           NaN   

                                         to_be_purchased  
sku                                                       
أرز الضحى مصري - 1 كيلو                              NaN  
أرز رفيع الأهرام - 1 كيلو                            NaN  
اريال جل - 2.5 لتر                                   NaN  
افيريدي ازرق AA - 20 حجر                             NaN  
افيريدي ازرق AAA - 40 حجر                            NaN  
...                                                  ...  
نسكافية بندق 3*1 ظرف (عرض) - 18 جرام                 NaN  
نسكافية شوكولاتة 3*1 ظرف (عرض) - 18 جرام             NaN  
نسكافية ظرف 3*1 (عرض) - 18 جرام                      NaN  
هوت صوص هاينز - 165 مل                               NaN  
هيد أند شولدرز اكياس 2+1 (عرض) - 5 مل                NaN  

[185 rows x 4 columns]

In [236]:
all_sales_data_SkuFor7Days['Sales_Count'] = sku_less_than_7days['Sales_Count']
all_sales_data_SkuFor7Days['Stock'] = sku_less_than_7days['Stock']
all_sales_data_SkuFor7Days['desired_stock'] = np.round((all_sales_data_SkuFor7Days['Sales_Count'] / 10) * 7)
all_sales_data_SkuFor7Days['to_be_purchased'] = np.round(all_sales_data_SkuFor7Days['desired_stock'] - all_sales_data_SkuFor7Days['Stock'])

In [237]:
all_sales_data_SkuFor7Days

Sales_Count  Stock  desired_stock  \
sku                                                                           
أرز الضحى مصري - 1 كيلو                           153     84          107.0   
أرز رفيع الأهرام - 1 كيلو                          50     25           35.0   
اريال جل - 2.5 لتر                                 18      2           13.0   
افيريدي ازرق AA - 20 حجر                           37     16           26.0   
افيريدي ازرق AAA - 40 حجر                          35     16           24.0   
...                                               ...    ...            ...   
نسكافية بندق 3*1 ظرف (عرض) - 18 جرام               97      9           68.0   
نسكافية شوكولاتة 3*1 ظرف (عرض) - 18 جرام           99     52           69.0   
نسكافية ظرف 3*1 (عرض) - 18 جرام                    90     43           63.0   
هوت صوص هاينز - 165 مل                             81     39           57.0   
هيد أند شولدرز اكياس 2+1 (عرض) - 5 مل              39      5           27.0   

                                          to_be_purchased  
sku                                                        
أرز الضحى مصري - 1 كيلو                              23.0  
أرز رفيع الأهرام - 1 كيلو                            10.0  
اريال جل - 2.5 لتر                                   11.0  
افيريدي ازرق AA - 20 حجر                             10.0  
افيريدي ازرق AAA - 40 حجر                             8.0  
...                                                   ...  
نسكافية بندق 3*1 ظرف (عرض) - 18 جرام                 59.0  
نسكافية شوكولاتة 3*1 ظرف (عرض) - 18 جرام             17.0  
نسكافية ظرف 3*1 (عرض) - 18 جرام                      20.0  
هوت صوص هاينز - 165 مل                               18.0  
هيد أند شولدرز اكياس 2+1 (عرض) - 5 مل                22.0  

[185 rows x 4 columns]